From 1ff3273833fa6c09e7a33ec54c8e6261ac04fcde Mon Sep 17 00:00:00 2001
From: Shuguang Gong <shuguang.gong@intel.com>
Date: Tue, 17 Mar 2015 15:30:12 +0800
Subject: [PATCH 381/462] ov2685: enable edge enhancement in sensor driver

Change-Id: I88182ea3478adebb10f360b7cbe1b4bd3d5caa05
Tracked-On:https://jira01.devtools.intel.com/browse/IMINAN-34925
Signed-off-by: Shuguang Gong <shuguang.gong@intel.com>
Reviewed-on: https://icggerrit.ir.intel.com/54141
Approver: Jing Tao <jing.tao@intel.com>
Build: IMIN (SC)
Integrator: WeiX Meng <weix.meng@intel.com>
Build: WeiX Meng <weix.meng@intel.com>
Reviewed-by: WeiX Meng <weix.meng@intel.com>
Maintainer: WeiX Meng <weix.meng@intel.com>
---
 .../camera/drivers/media/i2c/ov2685.c              | 22 ++++++++++++++++++++++
 .../camera/drivers/media/i2c/ov2685.h              |  1 +
 2 files changed, 23 insertions(+)

diff --git a/drivers/external_drivers/camera/drivers/media/i2c/ov2685.c b/drivers/external_drivers/camera/drivers/media/i2c/ov2685.c
index 7821b34..81b1c41 100644
--- a/drivers/external_drivers/camera/drivers/media/i2c/ov2685.c
+++ b/drivers/external_drivers/camera/drivers/media/i2c/ov2685.c
@@ -831,11 +831,33 @@ static int ov2685_get_intg_factor(struct i2c_client *client,
 	return 0;
 }
 
+ static int ov2685_s_sharpness(struct v4l2_subdev *sd, int *value)
+ {
+	struct i2c_client *client = v4l2_get_subdevdata(sd);
+	unsigned char values[] = {0x04, 0x0c, 0x10, 0x18, 0x20, 0x30, 0x3f};
+	int ret = 0;
+
+	if (*value > 3 || *value < -3)
+		return -EINVAL;
+
+	ret = ov2685_write_reg(client, OV2685_8BIT,
+		OV2685_REG_CIP_CTRL_0A, values[*value + 3]);
+
+	if (ret)
+		dev_err(&client->dev, "setting edge enhancement fails.\n");
+
+	return ret;
+}
+
 static long ov2685_ioctl(struct v4l2_subdev *sd, unsigned int cmd, void *arg)
 {
 	switch (cmd) {
 	case ATOMISP_IOC_S_EXPOSURE:
 		return ov2685_s_exposure(sd, arg);
+
+	case ATOMISP_IOC_S_SENSOR_EE_CONFIG:
+		return ov2685_s_sharpness(sd, arg);
+
 	default:
 		return -EINVAL;
 	}
diff --git a/drivers/external_drivers/camera/drivers/media/i2c/ov2685.h b/drivers/external_drivers/camera/drivers/media/i2c/ov2685.h
index ad33756..045176c 100644
--- a/drivers/external_drivers/camera/drivers/media/i2c/ov2685.h
+++ b/drivers/external_drivers/camera/drivers/media/i2c/ov2685.h
@@ -82,6 +82,7 @@
 #define OV2685_REG_EXPOSURE_1	0x3501
 #define OV2685_REG_EXPOSURE_2	0x3502
 #define OV2685_REG_EXPOSURE_AUTO	0x3503
+#define OV2685_REG_CIP_CTRL_0A	0x530a
 #define OV2685_AEC_CTRL0	0x3a00
 #define OV2685_AEC_CTRL2	0x3a02
 #define OV2685_REG_WPT		0x3a03
-- 
1.9.1

